src/common/fundFromTurbo.ts

```diff
@@ -0,0 +1,233 @@
+/**
+ * Copyright (C) 2022-2024 Permanent Data Solutions, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { ArconnectSigner, SignatureConfig } from '@dha-team/arbundles';
+import { AxiosInstance, RawAxiosRequestHeaders } from 'axios';
+import { v4 } from 'uuid';
+
+import {
+  AoMessageResult,
+  FundFromTurboSigner,
+  TransactionId,
+  WriteOptions,
+} from '../types/common.js';
+import { Intent } from '../types/io.js';
+import { mARIOToken } from '../types/token.js';
+import { toB64Url } from '../utils/base64.js';
+import { createAxiosInstance } from '../utils/http-client.js';
+import { ILogger, Logger } from './logger.js';
+
+export interface TurboConfig {
+  // The URL of the Turbo payment service
+  paymentUrl?: string;
+  // The logger to use
+  logger?: ILogger;
+  // The HTTP client to use
+  axios?: AxiosInstance;
+  signer?: FundFromTurboSigner;
+}
+
+export type InitiateArNSPurchaseParams = {
+  type?: string;
+  years?: number;
+  increaseQty?: number;
+  name: string;
+  intent: Intent;
+  processId?: TransactionId;
+};
+
+export async function signedRequestHeadersFromSigner(
+  signer: FundFromTurboSigner,
+  nonce: string = v4(),
+): Promise<RawAxiosRequestHeaders> {
+  await (signer as ArconnectSigner).setPublicKey?.();
+  const signature = await signer.sign(Uint8Array.from(Buffer.from(nonce)));
+
+  let publicKey: string;
+  switch (signer.signatureType) {
```

Comment:
we depend on Signer.signatureType for determining signer req approach rather than drilling `token` down and requiring integrators to provide `token` type